File: test_fixtures/sample.py
"""@toon
purpose: Sample Python fixture for testing the luny Python parser.
    This file contains various Python constructs including classes, functions,
    decorators, and type hints to verify extraction works correctly.

when-editing:
    - !Keep all export types represented for comprehensive testing
    - Maintain the mix of sync and async functions
    - Preserve decorator patterns

invariants:
    - All public items must have clear, testable names
    - Import statements should cover all supported patterns
    - Private functions (underscore prefix) should not be exported

do-not:
    - Remove any exports without updating corresponding tests
    - Use relative imports in this fixture

gotchas:
    - Python uses underscore convention for private items
    - Decorators affect function signatures
"""

from typing import Optional, List, Dict, Any, TypeVar, Generic
from dataclasses import dataclass, field
from abc import ABC, abstractmethod
from functools import wraps
import asyncio
import json
import os

# Type variables
T = TypeVar('T')
K = TypeVar('K')
V = TypeVar('V')


# Decorator
def log_calls(func):
    """Decorator that logs function calls."""
    @wraps(func)
    def wrapper(*args, **kwargs):
        print(f"Calling {func.__name__}")
        result = func(*args, **kwargs)
        print(f"{func.__name__} returned {result}")
        return result
    return wrapper


# Dataclass export
@dataclass
class UserConfig:
    """Configuration for a user."""
    id: str
    name: str
    email: Optional[str] = None
    settings: Dict[str, Any] = field(default_factory=dict)


# Abstract base class
class BaseService(ABC, Generic[T]):
    """Abstract base class for services."""

    @abstractmethod
    def get(self, id: str) -> Optional[T]:
        """Get an item by ID."""
        pass

    @abstractmethod
    def save(self, item: T) -> bool:
        """Save an item."""
        pass


# Concrete class
class UserService(BaseService[UserConfig]):
    """Service for managing users."""

    def __init__(self, data_dir: str = "data"):
        self.data_dir = data_dir
        self._cache: Dict[str, UserConfig] = {}

    def get(self, id: str) -> Optional[UserConfig]:
        """Get a user by ID."""
        if id in self._cache:
            return self._cache[id]

        filepath = os.path.join(self.data_dir, f"{id}.json")
        if os.path.exists(filepath):
            with open(filepath) as f:
                data = json.load(f)
                user = UserConfig(**data)
                self._cache[id] = user
                return user
        return None

    def save(self, user: UserConfig) -> bool:
        """Save a user to disk."""
        filepath = os.path.join(self.data_dir, f"{user.id}.json")
        try:
            with open(filepath, 'w') as f:
                json.dump(user.__dict__, f, indent=2)
            self._cache[user.id] = user
            return True
        except IOError:
            return False

    def _validate(self, user: UserConfig) -> bool:
        """Private validation method."""
        return len(user.id) > 0 and len(user.name) > 0


# Function exports
@log_calls
def validate_email(email: str) -> bool:
    """Validate an email address format."""
    return '@' in email and '.' in email.split('@')[1]


async def fetch_user_async(user_id: str) -> Optional[UserConfig]:
    """Async function to fetch a user."""
    await asyncio.sleep(0.1)  # Simulate network delay
    service = UserService()
    return service.get(user_id)


def create_user(name: str, email: Optional[str] = None) -> UserConfig:
    """Factory function to create a new user."""
    import uuid
    return UserConfig(
        id=str(uuid.uuid4()),
        name=name,
        email=email,
    )


# Constants
DEFAULT_TIMEOUT: int = 30
MAX_RETRIES: int = 3
VERSION: str = "1.0.0"


# Private function (should not be exported)
def _internal_helper() -> None:
    """Internal helper function."""
    pass


# Generic class
class Cache(Generic[K, V]):
    """Generic cache implementation."""

    def __init__(self, max_size: int = 100):
        self.max_size = max_size
        self._data: Dict[K, V] = {}

    def get(self, key: K) -> Optional[V]:
        return self._data.get(key)

    def set(self, key: K, value: V) -> None:
        if len(self._data) >= self.max_size:
            # Remove oldest entry
            oldest = next(iter(self._data))
            del self._data[oldest]
        self._data[key] = value

File: test_fixtures/sample.rb
=begin
@dose
purpose: Sample Ruby fixture for testing the luny Ruby parser.
    This file contains various Ruby constructs including classes, modules,
    methods, and metaprogramming patterns to verify extraction works correctly.

when-editing:
    - !Keep all export types represented for comprehensive testing
    - Maintain the mix of class and module patterns

invariants:
    - All public items must have clear, testable names
    - Include both instance and class methods

do-not:
    - Remove any exports without updating corresponding tests

gotchas:
    - Ruby private methods are declared after the `private` keyword
    - Module methods require explicit `module_function` or `self.` prefix
=end

require 'json'
require 'fileutils'
require_relative 'helper'

# Module with mixin functionality
module Loggable
  def log(message)
    puts "[#{Time.now.iso8601}] #{message}"
  end

  def log_error(message)
    log("ERROR: #{message}")
  end
end

# Module with class methods
module StringUtils
  module_function

  def camelize(str)
    str.split('_').map(&:capitalize).join
  end

  def underscore(str)
    str.gsub(/([A-Z])/, '_\1').downcase.sub(/^_/, '')
  end

  def truncate(str, length = 50)
    str.length > length ? "#{str[0...length]}..." : str
  end
end

# Configuration class
class UserConfig
  attr_accessor :id, :name, :email, :settings

  def initialize(id:, name:, email: nil, settings: {})
    @id = id
    @name = name
    @email = email
    @settings = settings
  end

  def to_h
    { id: @id, name: @name, email: @email, settings: @settings }
  end

  def to_json(*args)
    to_h.to_json(*args)
  end

  def self.from_json(json_str)
    data = JSON.parse(json_str, symbolize_names: true)
    new(**data)
  end
end

# Service class with inheritance
class BaseService
  include Loggable

  def initialize(data_dir = 'data')
    @data_dir = data_dir
    @cache = {}
  end

  def get(id)
    raise NotImplementedError, 'Subclasses must implement #get'
  end

  def save(item)
    raise NotImplementedError, 'Subclasses must implement #save'
  end

  protected

  def cache_get(id)
    @cache[id]
  end

  def cache_set(id, item)
    @cache[id] = item
  end
end

class UserService < BaseService
  def get(id)
    return cache_get(id) if cache_get(id)

    filepath = File.join(@data_dir, "#{id}.json")
    return nil unless File.exist?(filepath)

    json_data = File.read(filepath)
    user = UserConfig.from_json(json_data)
    cache_set(id, user)
    user
  rescue JSON::ParserError => e
    log_error("Failed to parse user file: #{e.message}")
    nil
  end

  def save(user)
    validate!(user)
    filepath = File.join(@data_dir, "#{user.id}.json")
    FileUtils.mkdir_p(@data_dir)
    File.write(filepath, user.to_json)
    cache_set(user.id, user)
    log("Saved user #{user.id}")
    true
  rescue IOError => e
    log_error("Failed to save user: #{e.message}")
    false
  end

  def all_users
    Dir.glob(File.join(@data_dir, '*.json')).map do |filepath|
      id = File.basename(filepath, '.json')
      get(id)
    end.compact
  end

  private

  def validate!(user)
    raise ArgumentError, 'User ID required' if user.id.nil? || user.id.empty?
    raise ArgumentError, 'User name required' if user.name.nil? || user.name.empty?
  end
end

# Factory module
module UserFactory
  def self.create(name:, email: nil)
    UserConfig.new(
      id: SecureRandom.uuid,
      name: name,
      email: email,
      settings: {}
    )
  end

  def self.create_admin(name:, email:)
    user = create(name: name, email: email)
    user.settings[:role] = 'admin'
    user
  end
end

# Constants
VERSION = '1.0.0'
DEFAULT_TIMEOUT = 30
MAX_RETRIES = 3
